### PR TITLE
error handling: edge cases

### PR DESCRIPTION
**Pull request by arnaudgeiser**
The `error-handling` branch contains a lot of changes we are not really comfortable with to safely integrate them on master. Still, this branch contains some unit tests that are valuable.

Instead of trying to rebase this branch on top of master, I decided to start from the unit tests and add the fixes for the tests that were failing. Here is how I end up with this PR.

As a reminder, here is what @kachayev intended to address :

 - both servers and clients close connection (context, to be more precise) in case of exceptions
 - error-logger parameter is added for HTTP client/server, Websocket client/server to customize exceptions logging for those exceptions that could not be visible to user's code in any other way (e.g. something that happens within HTTP server pipeline)
 - TimeoutException handlers are fixed in HTTP client (to avoid double disposal of the connections)
 -  HTTP decoder errors are now properly handled to deal with bad traffic, error code statuses are fixed
 - HTTP client propagates throwable to the caller whenever possible (rather than just relying on connection being closed later)
 - :aleph/interrupted? flag is introduce as a part of ring request object for HTTP server, making it's possible to track if streaming body was closed because of an error (e.g. decoder exception, premature close, etc)
 - a lot of unit tests to check failure scenarios

Some bits of those issues were fixed on various PR over last year.
This PR brings only some improvements on the following topics:

-  HTTP decoder errors are now properly handled to deal with bad traffic, error code statuses are fixed
 - HTTP client propagates throwable to the caller whenever possible (rather than just relying on connection being closed later)
 - a lot of unit tests to check failure scenarios

I would propose to merge this PR and close this one : https://github.com/clj-commons/aleph/pull/553

### CODE REVIEW

File: src/aleph/http/core.clj [hunk 1]
```diff
@@ -544,7 +553,7 @@
                 (try
                   (send-streaming-body ch msg body)
                   (catch Throwable e
-                    (log/error e "error sending body of type " class-name)
+                    (log/error "error sending body of type " class-name)
```
Comment (arnaudgeiser) [inline src/aleph/http/core.clj:556]: Since the exception is thrown to the client, we don't need to log anymore.

File: src/aleph/http/server.clj [hunk 2]
```diff
@@ -107,9 +107,7 @@
   (log/error e "error in HTTP handler")
   {:status 500
    :headers {"content-type" "text/plain"}
-   :body (let [w (java.io.StringWriter.)]
-           (.printStackTrace e (java.io.PrintWriter. w))
-           (str w))})
+   :body "Internal Server Error"})
```
Comment (arnaudgeiser) [inline src/aleph/http/server.clj:110]: It was pretty weird to return the details of the server internal error to the client (at least for security reasons)

File: src/aleph/http/core.clj [hunk 3]
```diff
@@ -305,15 +308,22 @@
...
+                      (comp
+                       (map
+                        (fn [x]
+                          (try
+                            (netty/to-byte-buf x)
+                            (catch Throwable e
+                              (log/error "error converting" (.getName (class x)) "to ByteBuf")
+                              (d/error! d e)
+                              (netty/close ch)
+                              nil))))
```
[the remark this answers is not in the document]
Comment (KingMob) [reply src/aleph/http/core.clj:325]: No. We have a `filter`, but that's not quite the same.

However, you can try using core `keep` in its transducer mode, and apply that. Streams work with transducers (and if not, that's a bug).

File: src/aleph/http/core.clj (hunk 1 above)
Comment (KingMob) [reply src/aleph/http/core.clj:556]: Why not? Client and server could be different parties, yes?

File: src/aleph/http/server.clj (hunk 2 above)
Comment (KingMob) [reply src/aleph/http/server.clj:110]: Good call. We may want to enable this with a configurable flag (if you trust your network), but it definitely shouldn't be the default.

**Comment by arnaudgeiser** [comment]
Rebased on master, no changes.

File: src/aleph/http/core.clj (hunk 1 above)
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: 312ffa88ad488c10ea26009dee8234e79585da9f

Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: Yes... reverted.

File: src/aleph/http/server.clj (hunk 2 above)
Comment (arnaudgeiser) [reply src/aleph/http/server.clj:110]: Yes, but let's not add this flag is not requested, right?

File: src/aleph/http/core.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: Good catch, thanks for the call.

81fde0056b388bee4915cc3ef671f415fba98840

File: src/aleph/http/server.clj (hunk 2 above)
Comment (KingMob) [reply src/aleph/http/server.clj:110]: Yeah, probably no need for that yet.

**Comment by arnaudgeiser** [comment]
On my side, I consider this ready.
Ignore the last commit as it's part of another PR for tests stabilization [1]

[1] : https://github.com/clj-commons/aleph/pull/661

**Comment by KingMob** [comment]
See my comment in the other PR about using port 0 for simpler, safer randomization.

File: src/aleph/http/client.clj [hunk 4]
```diff
@@ -143,21 +148,32 @@
...
+              (do
+                ;; flag that body was not completed succesfully
+                (d/success! @complete true)
+                (s/close! s))
+              (handle-decoder-failure ctx msg response-stream))
+            (let [content (.content ^HttpContent msg)]
+              (netty/put! (.channel ctx) @stream content)
+              (when (instance? LastHttpContent msg)
+                (d/success! @complete false)
+                (s/close! @stream))))
```
Comment (KingMob) [inline src/aleph/http/client.clj:176]: Are the `(d/success! @complete ...)` calls backwards?

It looks like you set `complete` to true when the decoder failed, and set it to false when it succeeded.

Comment (KingMob) [reply src/aleph/http/client.clj:176]: Hrmm, maybe not, I think I'd have to dig in to see how this is being used.

File: src/aleph/http/client.clj [hunk 5]
```diff
@@ -143,21 +148,32 @@
...
-              (s/close! @stream)))
+          (if (http/decoder-failed? msg)
+            ;; note that we are most likely to get this when dealing
+            ;; with transfer encoding chunked
+            (if-let [s @stream]
+              (do
+                ;; flag that body was not completed succesfully
+                (d/success! @complete true)
+                (s/close! s))
+              (handle-decoder-failure ctx msg response-stream))
```
Comment (KingMob) [inline src/aleph/http/client.clj:171]: It feels like we're duplicating a lot of logic here. Is there a way to simplify it? Maybe check for decoder failure first, and then have separate branches for the `Http*` types?

Comment (arnaudgeiser) [reply src/aleph/http/client.clj:171]: I will try to improve it.

File: src/aleph/http/client.clj (hunk 4 above)
Comment (arnaudgeiser) [reply src/aleph/http/client.clj:176]: `complete` is a little bit weird inside aleph.
It basically means __Did the HTTP request got completed before receiving the LastHtpContent?__.

`false` => everything is fine
`true`  => not completed successfully

Comment (arnaudgeiser) [reply src/aleph/http/client.clj:176]: And then, we end up here : https://github.com/clj-commons/aleph/blob/ce7e49e3c3eb0c10eae5e01b1fc579b318198c68/src/aleph/http.clj#L386-L389

File: src/aleph/http/client.clj (hunk 5 above)
Comment (arnaudgeiser) [reply src/aleph/http/client.clj]: https://github.com/clj-commons/aleph/pull/658/commits/6fc413ed4499b8c22aa29119503284410b142727

**Comment by arnaudgeiser** [comment]
Comments addressed.
Having a look at the diff might be simpler than looking at the last commit.

File: src/aleph/http/client.clj (hunk 4 above)
Comment (KingMob) [reply src/aleph/http/client.clj]: This is unintuitive, and if you can think of a better name, I'd support it.

File: src/aleph/http/client.clj [hunk 6]
```diff
@@ -195,6 +211,17 @@
...
+          (if (instance? HttpContent msg)
+            ;; note that we are most likely to get this when dealing
+            ;; with transfer encoding chunked
+            (if-let [s @stream]
+              (do
+                ;; flag that body was not completed succesfully
+                (d/success! @complete true)
+                (s/close! s))
+              (handle-decoder-failure ctx msg response-stream))
+            (handle-decoder-failure ctx msg response-stream))
```
Comment (KingMob) [inline src/aleph/http/client.clj:224]: Isn't this code identical to the other one? Maybe extract into a common fn?

File: src/aleph/http/client.clj (hunk 4 above)
Comment (arnaudgeiser) [reply src/aleph/http/client.clj]: Unfortunately, we are committed to that `aleph/complete` key on the response.
We are better off keeping it that way.

**Comment by arnaudgeiser** [comment]
I will rebase when the PR has been `Approved` (+ resolve conflicts and revert the changes to TCP/UDP tests)

File: src/aleph/http/core.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (KingMob) [reply src/aleph/http/core.clj]: Waitaminit...why was `keep` introduced, anyway? Did exceptions mean nils got on the stream? But it should have been closed...

[the remark this answers is not in the document]
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: (Sorry for the delay, I will come with an explanation later this week(-end?))

[the remark this answers is not in the document]
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: Actually, it's due to this Exception, but the tests should pass even with it :

```
java.lang.ClassCastException: class io.netty.channel.DefaultChannelPromise cannot be cast to class io.netty.buffer.ByteBuf (io.netty.channel.DefaultChannelPromise and io.netty.buffer.ByteBuf are in unnamed module of loader 'app')
	at aleph.http.core$send_streaming_body$fn__13303.invoke(core.clj:325)
	at aleph.netty.ChannelSink$fn__12725.invoke(netty.clj:429)
	at aleph.netty.ChannelSink.put(netty.clj:428)
	at manifold.stream.graph$async_send.invokeStatic(graph.clj:82)
	at manifold.stream.graph$async_send.invoke(graph.clj:78)
	at manifold.stream.graph$async_connect$this__10130.invoke(graph.clj:208)
	at manifold.stream.graph$async_connect$this__10130$fn__10131$fn__10132.invoke(graph.clj:189)
	at clojure.core$trampoline.invokeStatic(core.clj:6345)
	at clojure.core$trampoline.invoke(core.clj:6334)
	at manifold.stream.graph$async_connect$this__10130$fn__10131.invoke(graph.clj:189)
	at manifold.deferred.Listener.onSuccess(deferred.clj:220)
	at manifold.deferred.Deferred$fn__8571.invoke(deferred.clj:401)
	at manifold.deferred.Deferred.success(deferred.clj:401)
	at manifold.deferred$success_BANG_.invokeStatic(deferred.clj:246)
	at manifold.deferred$success_BANG_.invoke(deferred.clj:241)
	at manifold.stream.default.Stream$fn__10275.invoke(default.clj:173)
	at manifold.stream.default.Stream.put(default.clj:159)
	at manifold.stream.default.Stream.put(default.clj:195)
	at manifold.stream$map$fn__10859.invoke(stream.clj:621)
	at manifold.stream.Callback.put(stream.clj:454)
	at manifold.stream.graph$async_send.invokeStatic(graph.clj:82)
	at manifold.stream.graph$async_send.invoke(graph.clj:78)
	at manifold.stream.graph$sync_connect$f__8049__auto____10145.invoke(graph.clj:272)
	at clojure.lang.AFn.run(AFn.java:22)
	at io.aleph.dirigiste.Executor$3.run(Executor.java:318)
	at io.aleph.dirigiste.Executor$Worker$1.run(Executor.java:62)
	at manifold.executor$thread_factory$reify__7927$f__7928.invoke(executor.clj:70)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.lang.Thread.run(Thread.java:1589)
```

`netty/close` returns a `io.netty.channel.DefaultChannelPromise` that is put on the manifold stream that later is sent as a `DefaultChannelPromise` (hence the `keep` to remove it...).

[the remark this answers is not in the document]
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: I reverted my changes, but the current behavior seems fishy to me....

[the remark this answers is not in the document]
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: https://github.com/clj-commons/aleph/pull/658/commits/604a803e6b7728275c14ddf72d22478dc36fcd5c

[the remark this answers is not in the document]
Comment (arnaudgeiser) [reply src/aleph/http/core.clj]: Ok, tests are not passing on the CI when `keep`is not present.

=> https://app.circleci.com/pipelines/github/clj-commons/aleph/446/workflows/10a635c4-4ae9-48c8-995e-60a2f6a8d6c8/jobs/432

which requires this modification : https://github.com/clj-commons/aleph/pull/658/commits/80ae77fb0eb8b61bf87f59882976281b7bbdf9b6

[the remark this answers is not in the document]
Comment (KingMob) [reply src/aleph/http/core.clj]: That makes sense. We shouldn't be trying to convert a ChannelPromise to a ByteBuf, and really, we shouldn't be putting the ChannelPromise on the stream anyway. Stream closures (in the original sense of closing something) are handled out-of-band, we don't send terminal values or anything... (though I wonder if that might not be simpler).

**Comment by KingMob** [comment]
Merge when ready 👍🏻

**Pull request merged**